src/mixins/ui/lasso-tool-ui.js

```diff
@@ -246,23 +400,32 @@ class CircleShapeHandler extends ShapeHandler {
           shape.radius = Math.min(projDims[0], projDims[1])
         }
       }
-      // need to unset the active shape here first
-      // before calling setupFinalShape, because
-      // inside that call it's going to call destroy()
-      this.activeShape = null
       shape.setStyle(this.defaultStyle)
       this.setupFinalShape(shape)
     }
   }
 
-  destroy() {
-    if (this.activeShape) {
-      this.drawEngine.deleteShape(this.activeShape)
-      this.drawEngine.fire(LassoGlobalEventConstants.LASSO_SHAPE_DESTROY, {
-        shape: this.activeShape
-      })
-      this.activeShape = null
+  /**
+   * Called under various contexts to destroy either temporary shapes that might have
+   * been created during the shape draw and optionally destroy the primary shape that
+   * may have been drawn
+   * @param {DestroyType} destroy_type Enum describing the context under which destroy()
+   *                                   was called.
+   * @returns {Boolean} returrns true if one or more shapes were deleted as part of the destroy
+   */
+  destroy(destroy_type) {
+    let is_shape_deleted = false
+    if (DestroyType.isCancellableDestroyType(destroy_type)) {
```

Comment:
you only destroy primary shapes inside this `isCancellableDestroyType` check. All other derived `destroy` methods do something similar. Also returning whether shapes were destroyed or not. This indicates whether there was data to clear or not, indicating whether a draw-stopped signal should be thrown or not.